src/util/buf.rs

```diff
@@ -0,0 +1,560 @@
+    /// It's unsafe to read from the slice as the content is uninitialized.
+    pub unsafe fn with_len(len: usize) -> GrpcSlice {
+        GrpcSlice(grpc_slice_malloc(len))
+    }
+
+    /// Get the length of the data.
+    pub fn len(&self) -> usize {
+        unsafe {
+            if !self.0.refcount.is_null() {
+                self.0.data.refcounted.length
+            } else {
+                self.0.data.inlined.length as usize
+            }
+        }
+    }
+
+    /// Returns a slice of inner buffer.
+    pub fn as_slice(&self) -> &[u8] {
+        unsafe {
+            if !self.0.refcount.is_null() {
+                let start = self.0.data.refcounted.bytes;
+                let len = self.0.data.refcounted.length;
+                std::slice::from_raw_parts(start, len)
+            } else {
+                let len = self.0.data.inlined.length;
+                &self.0.data.inlined.bytes[..len as usize]
+            }
+        }
+    }
+
+    pub fn as_ptr(&self) -> *const u8 {
+        unsafe {
+            if !self.0.refcount.is_null() {
+                self.0.data.refcounted.bytes
+            } else {
+                self.0.data.inlined.bytes.as_ptr()
+            }
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+impl Clone for GrpcSlice {
+    /// Clone the slice.
+    ///
+    /// If the slice is not inlined, the reference count will be increased
+    /// instead of copy.
+    fn clone(&self) -> Self {
+        GrpcSlice(unsafe { grpc_slice_ref(self.0) })
+    }
+}
+
+impl Default for GrpcSlice {
+    /// Returns a default slice, which is empty.
+    fn default() -> Self {
+        GrpcSlice(unsafe { grpc_empty_slice() })
+    }
+}
+
+impl Debug for GrpcSlice {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        self.as_slice().fmt(f)
+    }
+}
+
+impl Drop for GrpcSlice {
+    fn drop(&mut self) {
+        unsafe {
+            grpc_slice_unref(self.0);
+        }
+    }
+}
+
+impl PartialEq<[u8]> for GrpcSlice {
+    fn eq(&self, r: &[u8]) -> bool {
+        // Technically, the equal function inside vtable should be used.
+        // But it's not cheap or safe to create a grpc_slice from rust slice.
+        self.as_slice() == r
+    }
+}
+
+impl PartialEq<GrpcSlice> for GrpcSlice {
+    fn eq(&self, r: &GrpcSlice) -> bool {
+        unsafe { grpc_slice_eq(self.0, r.0) != 0 }
+    }
+}
+
+/// `grpc_slice` use `grpc_slice_refcount` to trace the lifetime of inner data.
+///
+/// When ref count decreases to 0, there will be no access to the data.
+/// To utilize the mechanism, we need to define a struct which has
+/// `grpc_slice_refcount` type as the first field, so that it's safe to cast
+/// a pointer to the struct to `grpc_slice_refcount` and all ref/unref operations
+/// will be forwarded to our own implement.
+///
+/// Vec will be stored inside the struct so that it will be dropped automatically
+/// once the struct is dropped.
+#[repr(C)]
+struct VecRefCount {
```

Comment:
Could you document that `VecRefCount` must not be moved

Reply:
You could also add a `PhantomPinned` field so the compiler can help with enforcing that.

Reply:
Why? It can be moved around.